io.openems.edge.battery.fenecon.home/src/io/openems/edge/battery/fenecon/home/FeneconHomeBatteryImpl.java

```diff
@@ -294,111 +330,286 @@ protected ModbusProtocol defineModbusProtocol() throws OpenemsException {
+						String key = this.getSingleCellPrefix(j, i, t) + KEY_TEMPERATURE;
+						UnsignedWordElement uwe = new UnsignedWordElement(towerOffset + i * offset + tempOffset + j);
+						AbstractModbusElement<?> ame = m(this.channelMap.get(key).channelId(), uwe);
+						elements.add(ame);
+					}
+					this.getModbusProtocol().addTask(new FC3ReadRegistersTask(towerOffset + offset * i + tempOffset,
+							Priority.HIGH, elements.toArray(new AbstractModbusElement<?>[0])));
+				}
+			}
+		} catch (OpenemsException e) {
+			this.log.info("Cell Temperature and Voltage Dynamic Channels could not created");
+		}
+	}
+
+	/**
+	 * Gets the Number of Modules Per Tower.
+	 * 
+	 * @return the Number of Modules Per Tower as a {@link CompletableFuture}.
+	 * @throws OpenemsException on error
+	 */
+	private CompletableFuture<Integer> getNumberOfModulesPerTowers() {
+		final CompletableFuture<Integer> result = new CompletableFuture<Integer>();
+		try {
+			ModbusUtils.readELementOnce(this.getModbusProtocol(), new UnsignedWordElement(10024), true)
+					.thenAccept(numberOfModulesPerTower -> {
+						if (numberOfModulesPerTower == null) {
+							return;
+						}
+						result.complete(numberOfModulesPerTower);
+					});
+		} catch (OpenemsException e) {
+			this.log.info("Number of modules per tower could not read");
+		}
+		return result;
+	}
+
+	/**
+	 * Gets the Number of Towers.
+	 * 
+	 * @return the Number of Towers as a {@link CompletableFuture}.
+	 */
+	private CompletableFuture<Integer> getNumberOfTowers() {
+		final CompletableFuture<Integer> result = new CompletableFuture<Integer>();
+		try {
+			ModbusUtils.readELementOnce(this.getModbusProtocol(), new UnsignedWordElement(14000), true)
+					.thenAccept(softwareVersionOfTower3 -> {
+						if (softwareVersionOfTower3 == null) {
+							return;
+						}
+						if (softwareVersionOfTower3 != 0) {
+							// Three Towers available
+							result.complete(3);
+						} else {
+							try {
+								ModbusUtils
+										.readELementOnce(this.getModbusProtocol(), new UnsignedWordElement(12000), true)
+										.thenAccept(softwareVersionOfTower2 -> {
+											if (softwareVersionOfTower2 == null) {
+												return;
+											}
+											if (softwareVersionOfTower2 != 0) {
+												// Two Towers available
+												result.complete(2);
+											} else {
+												// One Tower available
+												result.complete(1);
+											}
+										});
+							} catch (OpenemsException e) {
+								this.log.info("softwareVersionOfTower2 could not detect ");
+							}
+						}
+					});
+		} catch (OpenemsException e) {
+			this.log.info("softwareVersionOfTower3 could not detect ");
+		}
+		return result;
+	}
+
+	private String getSingleCellPrefix(int num, int module, int tower) {
+		return "TOWER_" + tower + "_MODULE_" + module + "_CELL_" + String.format(NUMBER_FORMAT, num);
+	}
+
+	io.openems.edge.common.channel.ChannelId generateBcuChannel(int bcuNumber, String channelIdSuffix) {
```

Comment:
`generateBcuChannel` and `getSingleCellPrefix` sound very similar. Can you explain why you have both - or possibly merge them?